File: python/heap.py
from typing import TypeVar, Generic, List, Optional

T = TypeVar('T', int, str)

class MinHeap(Generic[T]):
    def __init__(self):
        self._heap: List[T] = []

    @property
    def size(self) -> int:
        return len(self._heap)

    @property
    def is_empty(self) -> bool:
        return len(self._heap) == 0

    def _get_parent_index(self, index: int) -> int:
        return (index - 1) // 2

    def _get_left_child_index(self, index: int) -> int:
        return 2 * index + 1

    def _get_right_child_index(self, index: int) -> int:
        return 2 * index + 2

    def _has_parent(self, index: int) -> bool:
        return self._get_parent_index(index) >= 0

    def _has_left_child(self, index: int) -> bool:
        return self._get_left_child_index(index) < len(self._heap)

    def _has_right_child(self, index: int) -> bool:
        return self._get_right_child_index(index) < len(self._heap)

    def _parent(self, index: int) -> T:
        return self._heap[self._get_parent_index(index)]

    def _left_child(self, index: int) -> T:
        return self._heap[self._get_left_child_index(index)]

    def _right_child(self, index: int) -> T:
        return self._heap[self._get_right_child_index(index)]

    def _swap(self, index_one: int, index_two: int) -> None:
        self._heap[index_one], self._heap[index_two] = self._heap[index_two], self._heap[index_one]

    def _heapify_up(self) -> None:
        index = len(self._heap) - 1
        while self._has_parent(index) and self._parent(index) > self._heap[index]:
            self._swap(self._get_parent_index(index), index)
            index = self._get_parent_index(index)

    def _heapify_down(self) -> None:
        index = 0
        while self._has_left_child(index):
            smaller_child_index = self._get_left_child_index(index)
            if self._has_right_child(index) and self._right_child(index) < self._left_child(index):
                smaller_child_index = self._get_right_child_index(index)

            if self._heap[index] < self._heap[smaller_child_index]:
                break
            else:
                self._swap(index, smaller_child_index)
            index = smaller_child_index

    def peek(self) -> Optional[T]:
        if self.is_empty:
            return None
        return self._heap[0]

    def poll(self) -> Optional[T]:
        if self.is_empty:
            return None
        item = self._heap[0]
        last_item = self._heap.pop()
        if self._heap:
            self._heap[0] = last_item
            self._heapify_down()
        return item

    def add(self, item: T) -> None:
        self._heap.append(item)
        self._heapify_up()

    def to_list(self) -> List[T]:
        return list(self._heap)

    def clear(self) -> None:
        self._heap = []

    def __str__(self) -> str:
        return str(self._heap)

class MaxHeap(Generic[T]):
    def __init__(self):
        self._heap: List[T] = []

    @property
    def size(self) -> int:
        return len(self._heap)

    @property
    def is_empty(self) -> bool:
        return len(self._heap) == 0

    def _get_parent_index(self, index: int) -> int:
        return (index - 1) // 2

    def _get_left_child_index(self, index: int) -> int:
        return 2 * index + 1

    def _get_right_child_index(self, index: int) -> int:
        return 2 * index + 2

    def _has_parent(self, index: int) -> bool:
        return self._get_parent_index(index) >= 0

    def _has_left_child(self, index: int) -> bool:
        return self._get_left_child_index(index) < len(self._heap)

    def _has_right_child(self, index: int) -> bool:
        return self._get_right_child_index(index) < len(self._heap)

    def _parent(self, index: int) -> T:
        return self._heap[self._get_parent_index(index)]

    def _left_child(self, index: int) -> T:
        return self._heap[self._get_left_child_index(index)]

    def _right_child(self, index: int) -> T:
        return self._heap[self._get_right_child_index(index)]

    def _swap(self, index_one: int, index_two: int) -> None:
        self._heap[index_one], self._heap[index_two] = self._heap[index_two], self._heap[index_one]

    def _heapify_up(self) -> None:
        index = len(self._heap) - 1
        while self._has_parent(index) and self._parent(index) < self._heap[index]:
            self._swap(self._get_parent_index(index), index)
            index = self._get_parent_index(index)

    def _heapify_down(self) -> None:
        index = 0
        while self._has_left_child(index):
            larger_child_index = self._get_left_child_index(index)
            if self._has_right_child(index) and self._right_child(index) > self._left_child(index):
                larger_child_index = self._get_right_child_index(index)

            if self._heap[index] > self._heap[larger_child_index]:
                break
            else:
                self._swap(index, larger_child_index)
            index = larger_child_index

    def peek(self) -> Optional[T]:
        if self.is_empty:
            return None
        return self._heap[0]

    def poll(self) -> Optional[T]:
        if self.is_empty:
            return None
        item = self._heap[0]
        last_item = self._heap.pop()
        if self._heap:
            self._heap[0] = last_item
            self._heapify_down()
        return item

    def add(self, item: T) -> None:
        self._heap.append(item)
        self._heapify_up()

    def to_list(self) -> List[T]:
        return list(self._heap)

    def clear(self) -> None:
        self._heap = []

    def __str__(self) -> str:
        return str(self._heap)

if __name__ == "__main__":
    print("Min Heap Example:")
    min_heap = MinHeap[int]()
    min_heap.add(5)
    min_heap.add(3)
    min_heap.add(7)
    min_heap.add(1)
    print(f"Min Heap: {min_heap}")
    print(f"Peek: {min_heap.peek()}")
    print(f"Poll: {min_heap.poll()}")
    print(f"Min Heap after poll: {min_heap}")

    print("\nMax Heap Example:")
    max_heap = MaxHeap[int]()
    max_heap.add(5)
    max_heap.add(3)
    max_heap.add(7)
    max_heap.add(1)
    print(f"Max Heap: {max_heap}")
    print(f"Peek: {max_heap.peek()}")
    print(f"Poll: {max_heap.poll()}")
    print(f"Max Heap after poll: {max_heap}")

File: rust/src/heap.rs
#![allow(dead_code)]
#![allow(unused_assignments)]

use std::fmt;
use std::cmp::PartialOrd;

pub struct MinHeap<T: Clone + PartialOrd> {
    heap: Vec<T>,
}

impl<T: Clone + PartialOrd + fmt::Display> MinHeap<T> {
    pub fn new() -> Self {
        MinHeap { heap: Vec::new() }
    }

    pub fn size(&self) -> usize {
        self.heap.len()
    }

    pub fn is_empty(&self) -> bool {
        self.heap.is_empty()
    }

    fn get_parent_index(&self, index: usize) -> usize {
        (index - 1) / 2
    }

    fn get_left_child_index(&self, index: usize) -> usize {
        2 * index + 1
    }

    fn get_right_child_index(&self, index: usize) -> usize {
        2 * index + 2
    }

    fn has_parent(&self, index: usize) -> bool {
        self.get_parent_index(index) < self.heap.len()
    }

    fn has_left_child(&self, index: usize) -> bool {
        self.get_left_child_index(index) < self.heap.len()
    }

    fn has_right_child(&self, index: usize) -> bool {
        self.get_right_child_index(index) < self.heap.len()
    }

    fn parent(&self, index: usize) -> &T {
        &self.heap[self.get_parent_index(index)]
    }

    fn left_child(&self, index: usize) -> &T {
        &self.heap[self.get_left_child_index(index)]
    }

    fn right_child(&self, index: usize) -> &T {
        &self.heap[self.get_right_child_index(index)]
    }

    fn swap(&mut self, index_one: usize, index_two: usize) {
        self.heap.swap(index_one, index_two);
    }

    fn heapify_up(&mut self) {
        let mut index = self.heap.len() - 1;
        while self.has_parent(index) && self.parent(index) > &self.heap[index] {
            let parent_index = self.get_parent_index(index);
            self.swap(parent_index, index);
            index = parent_index;
        }
    }

    fn heapify_down(&mut self) {
        let mut index = 0;
        while self.has_left_child(index) {
            let mut smaller_child_index = self.get_left_child_index(index);
            if self.has_right_child(index) && self.right_child(index) < self.left_child(index) {
                smaller_child_index = self.get_right_child_index(index);
            }

            if self.heap[index] < self.heap[smaller_child_index] {
                break;
            } else {
                self.swap(index, smaller_child_index);
            }
            index = smaller_child_index;
        }
    }

    pub fn peek(&self) -> Option<&T> {
        self.heap.first()
    }

    pub fn poll(&mut self) -> Option<T> {
        if self.is_empty() {
            return None;
        }
        let item = self.heap[0].clone();
        let last_item = self.heap.pop().unwrap();
        if !self.is_empty() {
            self.heap[0] = last_item;
            self.heapify_down();
        }
        Some(item)
    }

    pub fn add(&mut self, item: T) {
        self.heap.push(item);
        self.heapify_up();
    }

    pub fn to_vec(&self) -> Vec<T> {
        self.heap.clone()
    }

    pub fn clear(&mut self) {
        self.heap.clear();
    }
}

impl<T: Clone + PartialOrd + fmt::Display> fmt::Display for MinHeap<T> {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        let elements: Vec<String> = self.heap.iter().map(|x| x.to_string()).collect();
        write!(f, "[{}]", elements.join(", "))
    }
}

pub struct MaxHeap<T: Clone + PartialOrd> {
    heap: Vec<T>,
}

impl<T: Clone + PartialOrd + fmt::Display> MaxHeap<T> {
    pub fn new() -> Self {
        MaxHeap { heap: Vec::new() }
    }

    pub fn size(&self) -> usize {
        self.heap.len()
    }

    pub fn is_empty(&self) -> bool {
        self.heap.is_empty()
    }

    fn get_parent_index(&self, index: usize) -> usize {
        (index - 1) / 2
    }

    fn get_left_child_index(&self, index: usize) -> usize {
        2 * index + 1
    }

    fn get_right_child_index(&self, index: usize) -> usize {
        2 * index + 2
    }

    fn has_parent(&self, index: usize) -> bool {
        self.get_parent_index(index) < self.heap.len()
    }

    fn has_left_child(&self, index: usize) -> bool {
        self.get_left_child_index(index) < self.heap.len()
    }

    fn has_right_child(&self, index: usize) -> bool {
        self.get_right_child_index(index) < self.heap.len()
    }

    fn parent(&self, index: usize) -> &T {
        &self.heap[self.get_parent_index(index)]
    }

    fn left_child(&self, index: usize) -> &T {
        &self.heap[self.get_left_child_index(index)]
    }

    fn right_child(&self, index: usize) -> &T {
        &self.heap[self.get_right_child_index(index)]
    }

    fn swap(&mut self, index_one: usize, index_two: usize) {
        self.heap.swap(index_one, index_two);
    }

    fn heapify_up(&mut self) {
        let mut index = self.heap.len() - 1;
        while self.has_parent(index) && self.parent(index) < &self.heap[index] {
            let parent_index = self.get_parent_index(index);
            self.swap(parent_index, index);
            index = parent_index;
        }
    }

    fn heapify_down(&mut self) {
        let mut index = 0;
        while self.has_left_child(index) {
            let mut larger_child_index = self.get_left_child_index(index);
            if self.has_right_child(index) && self.right_child(index) > self.left_child(index) {
                larger_child_index = self.get_right_child_index(index);
            }

            if self.heap[index] > self.heap[larger_child_index] {
                break;
            } else {
                self.swap(index, larger_child_index);
            }
            index = larger_child_index;
        }
    }

    pub fn peek(&self) -> Option<&T> {
        self.heap.first()
    }

    pub fn poll(&mut self) -> Option<T> {
        if self.is_empty() {
            return None;
        }
        let item = self.heap[0].clone();
        let last_item = self.heap.pop().unwrap();
        if !self.is_empty() {
            self.heap[0] = last_item;
            self.heapify_down();
        }
        Some(item)
    }

    pub fn add(&mut self, item: T) {
        self.heap.push(item);
        self.heapify_up();
    }

    pub fn to_vec(&self) -> Vec<T> {
        self.heap.clone()
    }

    pub fn clear(&mut self) {
        self.heap.clear();
    }
}

impl<T: Clone + PartialOrd + fmt::Display> fmt::Display for MaxHeap<T> {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        let elements: Vec<String> = self.heap.iter().map(|x| x.to_string()).collect();
        write!(f, "[{}]", elements.join(", "))
    }
}

fn main() {
    println!("Min Heap Example:");
    let mut min_heap = MinHeap::<i32>::new();
    min_heap.add(5);
    min_heap.add(3);
    min_heap.add(7);
    min_heap.add(1);
    println!("Min Heap: {}", min_heap);
    if let Some(peek) = min_heap.peek() {
        println!("Peek: {}", peek);
    }
    if let Some(polled) = min_heap.poll() {
        println!("Polled: {}", polled);
    }
    println!("Min Heap after poll: {}", min_heap);

    println!("\nMax Heap Example:");
    let mut max_heap = MaxHeap::<i32>::new();
    max_heap.add(5);
    max_heap.add(3);
    max_heap.add(7);
    max_heap.add(1);
    println!("Max Heap: {}", max_heap);
    if let Some(peek) = max_heap.peek() {
        println!("Peek: {}", peek);
    }
    if let Some(polled) = max_heap.poll() {
        println!("Polled: {}", polled);
    }
    println!("Max Heap after poll: {}", max_heap);
}
